Add Origin-Bound Cookies (#4)
mikewest/cookie-incrementalism: draft-west-cookie-incrementalism.md

        request-host is identical to the cookie's domain.

    Or:

      - The cookie's host-only-flag is false, and the canonicalized
        request-host domain-matches the cookie's domain.

    * The request-uri's origin's port component port-matches the
      cookie.

    * The request-uri's origin's scheme component is identical to the
      cookie's scheme
~~~

This steps collects all cookies that could potentially be sent on this
request. Next add a new step that filters this list down to the cookies
that should actually be sent.

~~~
2. Let cookie-list be a new empty list.
   For each cookie in initial-cookie-list:
   1. Continue to the next cookie if cookie's host-only-flag is false
      and initial-cookie-list contains any cookies that meet all of
      the following criteria:
        * their host-only-flag is true.
        * their name matches cookie's name.
   2. Add cookie to cookie-list if it meets all of the following
      requirements:
        * The retrieval's URI's path path-matches the cookie's path.
        
        * If the cookie's http-only-flag is true, then exclude the
          cookie if the retrieval's type is "non-HTTP".
          

>         * If the cookie's same-site-flag is not "None" and the retrieval's same-site
>           status is "cross-site", then exclude the cookie unless all of the

Reviewer: ```suggestion
        * If the cookie's same-site-flag is not "None" and the HTTP request is
           cross-site (as defined in Section 5.2), then exclude the cookie unless all of the
```

          following conditions are met:

          * The retrieval's type is "HTTP".
          * The same-site-flag is "Lax" or "Default".
          * The HTTP request associated with the retrieval uses a "safe" method.
          * The target browsing context of the HTTP request associated with the

>             retrieval is a top-level browsing context.

Author: This feels clunky, but I wasn't sure how to better describe "Don't send an domain cookie if there is a matching origin cookie anywhere on the origin"

an opt-in mechanism. Helpfully there is already an attribute that
increases a cookie's scope: the `Domain` attribute. We can modify the
`Domain` attribute slightly such that it not only allows cookies to be
accessible by different hosts but also by different ports than the
origin's that set it. For convenience we'll call any cookie with a
`Domain` attribute a "domain cookie".

Example:

* A corporate network, `https://corp.example`, has various services
  each running on their own port. Those services share use of a token
  which allows users to use the services. To get a token a user logs
  in on `https://corp.example/login/` which creates the cookie
  `IDToken=a1b2bc3`. Next, to file an expense, the user needs to
  visit `https://corp.example:8443`, but when they do the IDToken
  isn't sent because the ports differ (443 vs 8443) and the user is
  denied access. To remedy this, the IDToken can have the `Domain`
  attribute added to it: `IDToken=a1b2d3; Domain=corp.example`.
  Now when the user visits `https://corp.example:8443` the token
  is sent and access is granted as expected.

Because any domain cookie is now exposed to multiple origins it means

> a cookie created by one origin can be overwritten by another origin. 

Author: Correct, it's meant to highlight that fact.

Example:

* `https://example.com:123` sets a cookie: `Set-Cookie:
  foo=domaincookie; Domain=example.com`. The user then visits
  `https://example.com:456` which sets the same cookie. This second
  cookie would match the first and then overwrite it.
    
Also, because domain cookies are less trusted due to their wider scope,
we'll want to avoid them shadowing non-domain cookies by disallowing
domain cookies from being sent if a matching non-domain cookie exists.
This is a departure from the status quo in which this shadowing
behavior is specifically allowed.

Example:

* `https://example.com:456` sets a non-domain cookie `Set-Cookie:
  foo=origincookie`. The user then visits `https://example.com:123`
  which sets a domain cookie `Set-Cookie: foo=domaincookie;
  Domain=example.com`. When the user returns to
  `https://example.com:456` only `foo=origincookie` is sent and the
  domain cookie is blocked from being sent. If the user were to then
  visit `https://sub.example.com:456` `foo=domaincookie` would be sent

>   since the domain would no longer be shadowing a non-domain cookie.

Reviewer: I mentioned `path` in some other thread that I can't find at the moment. I still think it's probably worth mentioning in prose somewhere. That is, consider `https://example.com:456/` setting `foo=origincookie; path=/whatever`. Does `foo=domaincookie` get sent on requests to paths other than`https://example.com:456/whatever`? I think it doesn't, based on my skim of the algorithm below, but spelling that out would be helpful.

Author: You're correct. I've added another example.

Finally, we don't ever want to allow a cookie to pass between schemes,
given the huge security differences between them. So there should be no
way for a server to specify that a given cookie should be sent to a
different scheme.

How we accomplish this is as follows:

First, add the concept of port matching which helps to simplify checking
if a cookie would match a port value. We can do that by adding a new

> section under 5.1 (this new section depends on the modification to 5.4

Reviewer: ```suggestion
section under 5.1 (this new section depends on the modification to 5.5
```

below)

~~~
5.1.5 Port matching 

An integer port-matches a given cookie if any of the following
conditions are true:
  1. The cookie's host-only-flag is false.

  2. The integer exactly matches the cookie's port value.
~~~

> Next, alter the storage model in Section 5.4 of {{RFC6265bis}} by

Reviewer: ```suggestion
Next, alter the storage model in Section 5.5 of {{RFC6265bis}} by
```

adding "scheme" and "port" to the list of fields the user agent stores
about each cookie, and setting them.

~~~

> 3.  Create a new cookie with name cookie-name, value cookie-value. Set

Reviewer: ```suggestion
5.  Create a new cookie with name cookie-name, value cookie-value. Set
```

    the creation-time and the last-access-time to the current date and
    time.
~~~

to: 

~~~
3.  Create a new cookie with name cookie-name, value cookie-value. Set
    the creation-time and the last-access-time to the current date and
    time, set the scheme to the request-uri's origin's scheme
    component, and set the port to the request-uri's origin's port
    component.
~~~

Because all cookies are now scoped by scheme it is unnecessary to
protect a secure scheme's cookies from an insecure scheme's cookies and

> thus we can remove step 13.

Reviewer: This would change behavior for sites mistakenly setting `secure` cookies on HTTP origins. Unless there's a good reason for changing that, it seems prudent to leave (admitedly ugly!) things like this alone.

Author: No good reason beyond it felt good to get rid of it. But you're right, we don't want to allow http to set `secure`.

To incorporate the new fields and "domain cookie" port matching, alter

> step 19 of the same algorithm from

Reviewer: ```suggestion
step 22 of the same algorithm from
```

       is at most one such cookie.)

    2. If the newly-created cookie was received from a "non-HTTP" API
       and the old-cookie's http-only-flag is true, abort these steps
       and ignore the newly created cookie entirely.

    3. Update the creation-time of the newly-created cookie to match
       the creation-time of the old-cookie.

    4. Remove the old-cookie from the cookie store.
~~~

to:

~~~
19. If the cookie store contains a cookie ("old-cookie") with the same
    name, scheme, domain, host-only-flag, path, and for which
    the old-cookie's port port-matches the newly-created cookie:

    1. If the newly-created cookie was received from a "non-HTTP" API
       and the old-cookie's http-only-flag is true, abort these steps
       and ignore the newly created cookie entirely.

    2. Update the creation-time of the newly-created cookie to match
       the creation-time of the old-cookie.

    3. Remove the old-cookie from the cookie store.
    (Notice that this algorithm maintains the invariant that there
    is at most one such cookie.)
~~~

With cookie storage taken care of now we move onto sending cookies. In

> section 5.5 alter step 1 from:

Reviewer: ```suggestion
section 5.6.1 alter step 1 from:
```

Also: step 1 contains more conditions than those listed here. It might be clearer to "replace step 1 with the following", and include all of them (`http-only-flag` and `same-site-flag`, at least, are currently missing).

Note: The request's URL when establishing a WebSockets connection has scheme "http" or "https", rather than "ws" or "wss". FETCH maps schemes when constructing the request. This mapping allows same-site cookies to be sent with WebSockets.

## Origin-Bound Cookies {#origin-bound-cookies}

Cookies are one of the few components of the web platform that are not

> scoped to the origin by default. The “Secure” attribute can scope a

Reviewer 2: nit: backticks (not quotes)

different scheme.

How we accomplish this is as follows:

First, add the concept of port matching which helps to simplify checking
if a cookie would match a port value. We can do that by adding a new
section under 5.1 (this new section depends on the modification to 5.4
below)

~~~
5.1.5 Port matching 

An integer port-matches a given cookie if any of the following
conditions are true:
  1. The integer exactly matches the cookie’s port value.

  2. The cookie’s host-only-flag is false.
~~~

We also want to add a sub-algorithm to protect origin cookies from being
overwritten/shadowed by domain cookies. This’ll be run during the
creation of new cookies in a later step.

~~~
5.1.6 New Cookie does not Shadow or Overwrite Origin-only Cookies

This algorithm protects any existing origin-only cookies from being
overwritten or shadowed by a new cookie ("new-cookie") with the Domain
attribute and is as follows:

If the new-cookie's host-only-flag is true, then return true.

If the new-cookie’s host-only-flag is false, then return false if the
cookie store contains at least one cookie ("equiv-cookie") that meets
all of the following criteria: 

>   1. Equiv-cookie’s name and scheme match new-cookie’s name and scheme.

Reviewer 2: nit: how about "are equal to" instead of "match" to avoid confusion with other notions of "match" defined above?

  2. Equiv-cookie’s host-only-flag is true.

  3. Equiv-cookie’s domain domain-matches the domain of new-cookie. 
  (Note: this is an asymmetrical comparison) 

Otherwise return true

Notice the asymmetric comparison during the domain-matching. It can be
helpful to think of domain-matching as: “A domain-matches B when A is
equivalent to or a sub-domain of B”. Since a cookie with the Domain
attribute is sent to all hosts that are equivalent to or a sub-domain
of its Domain’s value, this asymmetric comparison prevents a cookie
with the Domain attribute from being set when it would be sent to any
host that also contains an origin-only cookie.

Note: This comparison intentionally ignores the “path” component. The
intent is to protect a more tightly scoped origin bound cookie across
the entire origin.

Example: https://landing.site.example sets a cookie `Set-Cookie:
a=origincookie`. The user then visits https://account.site.example
which tries to set a cookie `Set-Cookie: a=domaincookie;
Domain=site.example`. The second cookie would fail to be set because

> the cookie would shadow“a=origincookie”.

Reviewer 2: nit: missing a space here after "shadow"

~~~

Next, alter the storage model in Section 5.4 of {{RFC6265bis}} by

> adding“scheme” and “port” to the list of fields the user agent stores

Reviewer 2: nit: missing a space after "adding"

> about each cookie, and setting them. Also, lets refer to the newly

Reviewer 2: nit: Word choice/punctuation with "lets"? Not sure if you meant "let's" as in "let us" or "lets" as in "this lets us".

"__Host-") can tighten that boundary, but these mechanisms are
little-used. Cookies that lack these protections easily move between
schemes (and even with these mechanisms will still move between ports).
The same cookies may be delivered to both the HTTP and HTTPS variants of
a given domain, even though the security properties of them differ
radically. As Section 8.6 of {{RFC6265bis}} points out, this gives
network attackers the ability to influence otherwise secured traffic by
modifying user state that flows to secure origins, and gives them
insight into user behavior, as securely-set cookies that lack the
`Secure` attribute likewise flow from secure origins to non-secure
origins. Additionally, cookies are delivered to all ports of a given
domain, which can't necessarily be trusted as Section 8.5 of
{{RFC6265bis}} points out. Similarly to above, attackers operating on a
different port could influence otherwise secured traffic.

We should remedy this defect by storing both a "scheme" and "port"
component along with the cookie, and use those components in cookies'
matching algorithms to ensure that a cookie is only sent to the origin

> that originally set it, thus keeping origins' state separate by default.

Author: I made a quick, partial, pass through your comments:
WDYT about linking to the [explainer](https://github.com/sbingler/Origin-Bound-Cookies) in order to off load the examples (this comment and others) there?

> that originally set it, thus keeping origins' state separate by default.

Reviewer: Linking out to the explainer is probably fine, as long as there's a clear story somewhere about what we're doing and why.

> that originally set it, thus keeping origins' state separate by default.

Author: I'm picking this back up again, I've gone through and mostly rewritten the opening section with a lot more examples. Let me know how it suits you now

It's possible that a site operator purposefully wants their cookies to
be usable across ports. For these situations it seems reasonable to
allow a server to specify that a given cookie should be accessible by
different origins. Since the Domain attribute already does this, by
matching multiple hosts, we can extend it to also match all ports.
Thereby allowing a cookie with the Domain attribute, a "domain cookie",

> to be accessible from any port value. This behavior extends to

Author: How's that?

overwriting cookies as well. Examples:

1. https://example.com:123 sets a cookie: `Set-Cookie: a=domaincookie;
    Domain=example.com`. The user then visits https://example.com:456
    which sets the same cookie. This second cookie would match the
    first and then overwrite it. 
2. https://example.com:123 sets a cookie `Set-Cookie: a=domaincookie;
    Domain=example.com`. The user then visits https://example.com:456
    which sets `Set-Cookie: a=origincookie`. The second cookie would

>     match the first and overwrite it.

Author: How's the new change?

Given that these "domain cookies" are less trusted (due to their
accessibility by different hosts and ports) we should try and prevent
them from shadowing "origin cookies". Example:

1. https://example.com:456 sets a cookie `Set-Cookie: a=origincookie`.
   The user then visits https://example.com:123 which tries to set a
   cookie `Set-Cookie: a=domaincookie; Domain=example.com`. The second
   cookie would fail to be set because the cookie it's attempting to
   overwrite is an origin cookie.

Finally, we don't ever want to allow a cookie to pass between schemes,
given the huge security differences between them. So there should be no
way for a server to specify that a given cookie should be sent to a
different scheme.

How we accomplish this is as follows:

First, add the concept of port matching which helps to simplify checking
if a cookie would match a port value. We can do that by adding a new
section under 5.1 (this new section depends on the modification to 5.4

> below)

Author: Are you recommending to split up the introduction of these concepts within the explanation or to ship the feature in two distinct parts?

> below)

Reviewer: It's been a little while, but I think I was suggesting treating the `scheme` and the `port` changes as distinct phases that you could ship along different timelines.

> below)

Reviewer: (Reading through this again, I still think these changes would be simpler to introduce as distinct alterations. Perhaps it would make sense to chat through your deployment story, as I might well be missing something that makes it prudent to ship all this as one change.)

> below)

Author: Doing as part of one proposal was primarily for three reasons:

- The story of aligning cookies on the origin, a well known concept (`Domain` not withstanding).
- The working assumption is that binding to scheme will be harder to ship, and that adding port onto that isn't much more work
  - Metrics also indicate that restriction the port is likely to be less breaking than scheme
- Avoiding annoying users more than necessary via death by 1000 changes.

> below)

Reviewer: I was thinking about this in the other direction: `port` seems much easier to ship, so we shouldn't block it on `scheme`. :)

> below)

Author: Spoke offline, I'm still considering shipping the feature separately but I won't block the spec change work on that.

~~~
5.1.5 Port matching 

An integer port-matches a given cookie if any of the following
conditions are true:
  1. The integer exactly matches the cookie's port value.

  2. The cookie's host-only-flag is false.
~~~

We also want to add a sub-algorithm to protect origin cookies from being
overwritten/shadowed by domain cookies. This'll be run during the
creation of new cookies in a later step.

~~~
5.1.6 New Cookie does not Shadow or Overwrite Origin-only Cookies

This algorithm protects any existing origin-only cookies from being
overwritten or shadowed by a new cookie ("new-cookie") with the Domain
attribute and is as follows:

If the new-cookie's host-only-flag is true, then return true.

If the new-cookie's host-only-flag is false, then return false if the
cookie store contains at least one cookie ("equiv-cookie") that meets
all of the following criteria: 
  1. Equiv-cookie's name and scheme are equal to new-cookie's name and
     scheme.

  2. Equiv-cookie's host-only-flag is true.

  3. Equiv-cookie's domain domain-matches the domain of new-cookie. 
  (Note: this is an asymmetrical comparison) 

> Otherwise return true

Author: I like it, changed.

Notice the asymmetric comparison during the domain-matching. It can be
helpful to think of domain-matching as: "A domain-matches B when A is
equivalent to or a sub-domain of B". Since a cookie with the Domain
attribute is sent to all hosts that are equivalent to or a sub-domain
of its Domain's value, this asymmetric comparison prevents a cookie
with the Domain attribute from being set when it would be sent to any

> host that also contains an origin-only cookie.

Author: Interesting, are you thinking of a situation in which:

1. `separated-for-security.google.com` sets `data=not_imporant` as an origin cookie.

2. Some `*.google.com` page tries setting a cookie `data=important; Domain=google.com` with eTLD+1 scope but the operation fails because it would shadow the previous?

> host that also contains an origin-only cookie.

Reviewer: Right. This is the scenario I was concerned about. I haven't done any analysis, but this seems like a risky extension in the absence of data.

> host that also contains an origin-only cookie.

Author: As a bit of a non-answer, I think the correct thing to do would be to separate security boundaries like that by eTLD+1. We already have the problem where a malicious sub-domain can mess with cookies of other origins. Although the other origins could always overwrite those and wouldn't be completely locked out...

Let's just deprecate `Domain` :)

Realistically I do see the issue, perhaps softening the policy so that domain cookies aren't sent when a corresponding origin cookie exists would be better? This way other origins can still create them, but we wouldn't have to worry about them shadowing.
So:
`*.google.com` would receive `data=important`
`separated-for-security.google.com` would only get `data=not_imporant`
`sub.separated-for-security.google.com` would get `data=important`

> host that also contains an origin-only cookie.

Reviewer: > I think the correct thing to do would be to separate security boundaries like that by eTLD+1

I mean, even more correctly, we should only have origin-bound tokens ("just deprecate `Domain`", as you note below: state tokens!). But, folks use same-site-different-origin domains today as security boundaries for all the things in the platform that _are_ origin-bound, while retaining cross-origin-same-site cookies for persistent authentication. We can (and, in the long term, should!) ask them to switch to origin-bound state through an SSO-like infrastructure, but that's complicated.

> `*.google.com` would receive `data=important`
> `separated-for-security.google.com` would only get `data=not_imporant`
> `sub.separated-for-security.google.com` would get `data=important`

As we've noted elsewhere, this is my preferred resolution. It will require you to do some work with regard to the other pieces that affect ordering in the `cookie` header to make sure we're picking the right one (`path` certainly), but this approach seems least-likely to accidentally break things.

> host that also contains an origin-only cookie.

Author: Sounds like a plan, I'll rework the change.

> path certainly
Not sure I follow, what do you mean?

> host that also contains an origin-only cookie.

Author: I've updated the processing in order to allow `Domain` cookies to always be set and to block them from being accessed when there is a matching origin cookie.  Let me know what you think.

(Your cleanup comments are still on my radar, I just want to get this hashed out first)

> host that also contains an origin-only cookie.

Author: Friendly ping

Note: The request's URL when establishing a WebSockets connection has scheme "http" or "https", rather than "ws" or "wss". FETCH maps schemes when constructing the request. This mapping allows same-site cookies to be sent with WebSockets.

## Origin-Bound Cookies {#origin-bound-cookies}

Cookies are one of the few components of the web platform that are not
scoped to the origin by default. This difference in scoping means that
cookies have weaken confidentiality and integrity compared with other
storage APIs on the web platform.

Examples:

1. `https://somesite.com` sets a simple cookie, `secret=123456`,

>     which contains private information about a user. Information that

Reviewer: Style nit: The body of each of these items is offset from the first line by one character. Is that intentional?

    an attacker wishes to learn. To do so the attacker
    man-in-the-middles the user, and then tricks them into visiting
    `http://somesite.com` (note the insecure scheme). When the user
    visits that page their browser will send the `secret` cookie and
    the attacker can see it.

2. Similarly, if the attacker has somehow compromised a service running
    on a different port on the same server, let's say port 345, as
    `https://somesite.com` then they could trick the user into visiting
    `https://somesite.com:345`, the user's browser will send the `secret`
    cookie, and once again the attacker can see it.

Even more, through the same techniques, an attacker can also modify
a user's cookies, sending a `Set-Cookie` field instead of simply
eavesdropping.

All of these examples are possible because cookies by default do not
care about the scheme or port of their connection. As long as the host
matches the cookie will be accessible.

Some of these shortcomings can be alleviated: the `Secure` attribute
scopes a cookie to only be accesible on secure schemes and the cookie

> prefixes, `__Secure-` and `__Host-`, can tighten that boundary more.

Reviewer: Pedantic Nit: The prefixes don't tighten the boundary, they prevent the server from accidentally loosening it. The lack of a `Domain` attribute is what makes `__Host-` cookies more tightly scoped.

Author: True, I've reworded them to keep the focus on the "secure" attribute.

While these solve the problem shown in the first example, they are all
opt-in and are not always used. They also do nothing to help solve

> the port problem shown in the example two. In fact, there are currently

Reviewer: Style nit: 

```suggestion
the port problem shown in the second example. In fact, there are currently
```

   visit `https://corp.example:8443`, but when they do the IDToken
   isn't sent because the ports differ (443 vs 8443) and the user is
   denied access. To remedy this, the IDToken can have the `Domain` 
   attribute added to it: `IDToken=a1b2d3; Domain=corp.example`.
   Now when the user visits `https://corp.example:8443` the token
   is sent and access is granted as expected.

Because any domain cookie is now exposed to multiple origins it means
a cookie created by one origin can be overwritten by another origin. 

Examples:

1. `https://example.com:123` sets a cookie: `Set-Cookie: 
    foo=domaincookie; Domain=example.com`. The user then visits 
    `https://example.com:456` which sets the same cookie. This second
    cookie would match the first and then overwrite it. 
2. `https://example.com:123` sets a cookie 
    `Set-Cookie: foo=domaincookie; Domain=example.com`. The user then
    visits `https://example.com:456` which sets `Set-Cookie: 
    foo=origincookie`. The second cookie would match the first an
    overwrite it.
    
Also, because domain cookies are less trusted due to their wider scope,
we'll want to avoid allowing them to shadow non-domain cookies by

> disallowing domain cookies from co-exiting or overwriting a matching

Reviewer: Nit: "co-exiting" seems like a typo (or a way of describing multiple cookies sent in the same outgoing request that I haven't heard before?) Perhaps "co-existing with"?

non-domain cookie. This is a departure from the status quo in which

> this shadowing behavior is specifically allowed.

Reviewer: I haven't looked at the algorithms below yet, but it's unclear to me what "overwrite" means in this context.

You're generally suggesting that cookies should be origin-bound, and shouldn't affect other origins' state. That can happen today in a few scenarios like the ones you spell out above, and I agree that it would be unfortunate if `https://sub.example.com:12345`'s `Domain=example.com` cookie could shadow `https://super-secure.example.com/`'s origin-bound cookie. By the same token, though, it seems bad for the latter to be able to remove state from the former by "overwriting" its cookie.

Another approach would store both cookies, but send only the most specific cookie for a given request in a case where there's both an origin-bound cookie and a domain cookie that match a given endpoint. Maybe this is what you mean!

Author: > but it's unclear to me what "overwrite" means in this context.

This proposal removes the allowance for cookies to differ only in the host only flag. I.e.: `foo=bar` and `foo=bar; Domain=example.com` would be explicitly disallowed.

Since only a single version of a cookie can exist "overwriting" refers to the domain cookies deleting the origin cookie and being set itself.

> Maybe this is what you mean!

I wish I could take credit for that, but looking at my design doc it seems I actually wanted to prevent the cookie from being set in the first place. That was over a year ago now and I can no longer recall why I thought that was preferable.
I recommended the same thing in a response to a different comment, so it does seem we're in agreement on this.

Reviewer: Addressed in the [thread above](https://github.com/mikewest/cookie-incrementalism/pull/4#discussion_r858363973). I'll resolve this comment.

Example:

> * `https://example.com:456` sets a non-domain cookie`Set-Cookie:

Reviewer: ```suggestion
* `https://example.com:456` sets a non-domain cookie `Set-Cookie:
```

   foo=origincookie`. The user then visits `https://example.com:123`
   which tries to set a domain cookie `Set-Cookie: foo=domaincookie;
   Domain=example.com` This set fails because a domain cookie should

>    never overwrite a matching non-domain cookie.

Reviewer: Doesn't this leak state from `https://example.com:456` to `https://example.com:123`? It seems better to me to store both cookies, as above, and to modify the retrieval algorithm in 5.6.3 accordingly.

Finally, we don't ever want to allow a cookie to pass between schemes,
given the huge security differences between them. So there should be no
way for a server to specify that a given cookie should be sent to a

> different scheme.

Reviewer: As in one of the other comment threads, this is going to be a big change in itself, and it might make sense for you to structure the proposal as explicitly staged.

> How we accomplish this is as follows:

Reviewer: Style nit: "We accomplish this as follows:"
